### PR TITLE
Add deployments autodiscovery

### PR DESCRIPTION
**Pull request by hamadise**
Add statefulset autodiscovery
Add HPA autodiscovery
Add deployment autoscaling without HPA
Add deployment autoscaling with HPA

### CODE REVIEW

File: pkg/provider/provider.go [hunk 1]
```diff
@@ -59,6 +62,9 @@ type TestEnvironment struct { // rename this with testTarget
 	MultusIPs          map[*v1.Pod]map[string][]string
 	SkipNetTests       map[*v1.Pod]bool
 	SkipMultusNetTests map[*v1.Pod]bool
+	Deployments        []*v1apps.Deployment
+	SatetfulSets       []*v1apps.StatefulSet
+	HorizentalScaller  map[string]*v1scaling.HorizontalPodAutoscaler
```
Comment (edcdavid) [inline pkg/provider/provider.go:67]: typo HorizentalScaller->HorizontalScaler

File: pkg/autodiscover/autodiscover_podset.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+		}
+	}
+	if len(statefulset) == 0 {
+		logrus.Info("did not find any statefulset")
+	}
+	return statefulset
+}
+
+func findHpaControllers(cs *kubernetes.Clientset, namespaces []string) map[string]*v1scaling.HorizontalPodAutoscaler {
+	m := make(map[string]*v1scaling.HorizontalPodAutoscaler)
```
Comment (edcdavid) [inline pkg/autodiscover/autodiscover_podset.go:104]: Shouldn't we also label the hpas under test?

File: cnf-certification-test/lifecycle/suite.go [hunk 3]
```diff
@@ -142,3 +152,31 @@ func testPodsOwnerReference(env *provider.TestEnvironment) {
...
+		ginkgo.By("Testing deployment scaling")
+		defer env.SetNeedsRefresh()
+
+		if len(env.Deployments) == 0 {
+			ginkgo.Skip("No test deployments found.")
+		}
+		failedDeployments := []string{}
+		skippedDeployments := []string{}
+		for i := range env.Deployments {
+			if !scaling.TestDeploymentScaling(env.Deployments[i], env.HorizentalScaller, timeout) {
```
Comment (edcdavid) [inline cnf-certification-test/lifecycle/suite.go:168]: for clarity, suggesting expanding the TestDeploymentScaling function here (e.g. delete the function). and add a comment explaining that if a matching hpa is present, scaling is performed using the hpa, and if not, it is performed using deployments directly

File: pkg/autodiscover/autodiscover_podset.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+	for _, ns := range namespaces {
+		for _, l := range labels {
+			label := buildLabelQuery(l)
+			logrus.Trace("find StatefulSet in ", ns, " using label ", label)
+			options := metav1.ListOptions{}
+			options.LabelSelector = label
+			statefulSetClient := appClient.StatefulSets(ns)
+			dps, err := statefulSetClient.List(context.TODO(), options)
+			if err != nil {
+				logrus.Errorln("error when listing StatefulSets in ns=", ns, " label=", label, " try to proceed")
```
Comment (edcdavid) [inline pkg/autodiscover/autodiscover_podset.go:91]: try -> trying

File: cnf-certification-test/lifecycle/scaling/scaling.go [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+		}
+		hpa.Spec.MinReplicas = &min
+		hpa.Spec.MaxReplicas = max
+		_, err = clients.K8sClient.AutoscalingV1().HorizontalPodAutoscalers(namespace).Update(context.TODO(), hpa, v1machinery.UpdateOptions{})
+		if err != nil {
+			logrus.Error("can't Update autoscaler to scale ", namespace, ":", deploymentName, " error=", err)
+			return err
+		}
+		ready := isDeploymentReady(namespace, deploymentName, timeout)
+		if !ready {
```
Comment (sebrandon1) [inline cnf-certification-test/lifecycle/scaling/scaling.go:236]:
```suggestion
		if !isDeploymentReady(namespace, deploymentName, timeout) {
```

File: cnf-certification-test/lifecycle/scaling/scaling.go [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+			return err
+		}
+		dp.Spec.Replicas = &replicas
+		_, err = clients.AppsClients.Deployments(namespace).Update(context.TODO(), dp, v1machinery.UpdateOptions{})
+		if err != nil {
+			logrus.Error("can't update deployment ", namespace, ":", name)
+			return err
+		}
+		ready := isDeploymentReady(namespace, name, timeout)
+		if !ready {
```
Comment (sebrandon1) [inline cnf-certification-test/lifecycle/scaling/scaling.go:140]:
```suggestion
		if !isDeploymentReady(namespace, name, timeout) {
```

**Review by sebrandon1 — COMMENTED** [review]
Just some things I found.  Looks good overall.

File: pkg/autodiscover/autodiscover_podset.go (hunk 4 above)
Comment (hamadise) [reply pkg/autodiscover/autodiscover_podset.go]: done

File: cnf-certification-test/lifecycle/suite.go (hunk 3 above)
Comment (hamadise) [reply cnf-certification-test/lifecycle/suite.go]: done

File: pkg/provider/provider.go [hunk 7]
```diff
@@ -124,21 +133,26 @@ func BuildTestEnvironment() { //nolint:funlen
...
 	env.DebugPods = make(map[string]*v1.Pod)
-	for i := 0; i < len(debugPods); i++ {
-		nodeName := debugPods[i].Spec.NodeName
-		env.DebugPods[nodeName] = &debugPods[i]
+	for i := 0; i < len(data.DebugPods); i++ {
+		nodeName := data.DebugPods[i].Spec.NodeName
+		env.DebugPods[nodeName] = &data.DebugPods[i]
+	}
+	for i := range data.Csvs {
+		env.Csvs = append(env.Csvs, &data.Csvs[i])
```
[the remark this answers is not in the document]
Comment (hamadise) [reply pkg/provider/provider.go:142]: we want pointers not the object themselves.

File: cnf-certification-test/lifecycle/scaling/scaling.go (hunk 5 above)
Comment (hamadise) [reply cnf-certification-test/lifecycle/scaling/scaling.go]: fixed.

File: pkg/provider/provider.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (sebrandon1) [reply pkg/provider/provider.go:142]: Yup, that's what I figured.  Disregard this.  😄

File: cnf-certification-test/lifecycle/scaling/scaling.go [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+	if retryErr != nil {
+		logrus.Error("can't scale hpa ", namespace, ":", hpaName, " error=", retryErr)
+		return false
+	}
+	return true
+}
+
+// TestDeploymentScaling test scaling of deployment
+// This is the entry point for deployment scaling tests
+func TestDeploymentScaling(deployment *v1app.Deployment, hpas map[string]*v1autoscaling.HorizontalPodAutoscaler, timeout time.Duration) (failed bool) {
```
Comment (edcdavid) [inline cnf-certification-test/lifecycle/scaling/scaling.go:249]: Remove dead code

Comment (hamadise) [reply cnf-certification-test/lifecycle/scaling/scaling.go]: fixed

File: cnf-certification-test/lifecycle/suite.go [hunk 9]
```diff
@@ -169,3 +178,46 @@ func testPodNodeSelectorAndAffinityBestPractices(env *provider.TestEnvironment)
...
+	testID := identifiers.XformToGinkgoItIdentifier(identifiers.TestDeploymentScalingIdentifier)
+	ginkgo.It(testID, ginkgo.Label(testID), func() {
+		ginkgo.By("Testing deployment scaling")
+		defer env.SetNeedsRefresh()
+
+		if len(env.Deployments) == 0 {
+			ginkgo.Skip("No test deployments found.")
+		}
+		failedDeployments := []string{}
+		skippedDeployments := []string{}
```
Comment (edcdavid) [inline cnf-certification-test/lifecycle/suite.go:191]: seems like skippedDeployments is not used

**Review by edcdavid — APPROVED** [review]

**Review by sebrandon1 — APPROVED** [review]

**Pull request merged**